### ISSUE TITLE
Issues building on NERSC's Cori System

### ISSUE THREAD

**Issue by JBlaschke** [block-hczhai/block2-preview#11]
Hi,

I am trying to build on NERSC's Cori Cray-XC system on behalf of a user. I have tried to build using the Intel and GNU compilers, getting different errors (detailed below). 

The TL;DR is that:
1. If compiling with Intel, then `icpc` consumes too many resources for our login nodes and is killed.
2. If compiling with GNU, the compiler complains about a syntax error.

Please advise how to compile on Cray systems, eg. which compiler do you recommend. You can find the detailed errors and CMake output below.

## Using Intel

```
zhcui@cori02:/global/cscratch1/sd/zhcui/.consulting/block2-preview/build>  CC=cc CXX=CC cmake .. -DUSE_MKL=ON -DBUILD_LIB=ON -DMPI=ON -DOMP_LIB=GNU -DLARGE_BOND=ON -DUSE_SG=ON -DUSE_BIG_SITE=ON -DCMAKE_CXX_FLAGS=-qoverride-limits
-- The C compiler identification is Intel 19.1.2.20200623
-- The CXX compiler identification is Intel 19.1.2.20200623
-- Cray Programming Environment 2.7.10 C
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /opt/cray/pe/craype/2.7.10/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Cray Programming Environment 2.7.10 CXX
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/cray/pe/craype/2.7.10/bin/CC - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/bin/python3 (found version "3.9.7")
-- Found PythonLibs: /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/lib/libpython3.so (found suitable version "3.9.7", minimum required is "3.9")
-- PROJECT_NAME = block2
-- PYTHON_VERSION_MAJOR = 3
-- PYTHON_VERSION_MINOR = 9
-- PYTHON_LIBRARIES = /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/lib/libpython3.so
-- PYTHON_EXECUTABLE = /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/bin/python3
-- PYTHON_EXECUTABLE_HINT =
-- PYTHON_INCLUDE_DIRS = /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/include/python3.9
-- PYLIB_SUFFIX = .cpython-39-x86_64-linux-gnu.so
-- PYBIND_INCLUDE_DIRS = /global/homes/z/zhcui/.local/cori/3.9-anaconda-2021.11/lib/python3.9/site-packages/pybind11/include
-- BUILD_LIB = ON
-- MKL_INCLUDE_DIR = /opt/intel/compilers_and_libraries_2020.2.254/linux/mkl/include
-- MKL_LIBS = -Wl,--no-as-needed;/global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/lib/libiomp5.so;/opt/intel/compilers_and_libraries_2020.2.254/linux/mkl/lib/
intel64/libmkl_intel_lp64.so;/opt/intel/compilers_and_libraries_2020.2.254/linux/mkl/lib/intel64/libmkl_core.so;/opt/intel/compilers_and_libraries_2020.2.254/linux/mkl/lib/intel64/li
bmkl_intel_thread.so;/opt/intel/compilers_and_libraries_2020.2.254/linux/mkl/lib/intel64/libmkl_avx2.so;/opt/intel/compilers_and_libraries_2020.2.254/linux/mkl/lib/intel64/libmkl_avx
512.so
-- Found MPI_C: /opt/cray/pe/craype/2.7.10/bin/cc (found version "3.1")
-- Found MPI_CXX: /opt/cray/pe/craype/2.7.10/bin/CC (found version "3.1")
-- Found MPI: TRUE (found version "3.1")
-- MPI_COMPILE_FLAGS =
-- MPI_LINK_FLAGS =
-- MPI_INCLUDE_PATH =
-- MPI_CXX_LIBRARIES =
-- OPT_FLAG = -O3;-funroll-loops;-qopenmp;-Werror;-Werror=return-type;-fvisibility=hidden;-Wno-error=attributes;-Wno-attributes
-- BOND_FLAG = -D_LARGE_BOND
-- MKL_FLAG = -D_HAS_INTEL_MKL=2
-- CORE_FLAG = -D_USE_CORE
-- DMRG_FLAG = -D_USE_DMRG
-- BIG_SITE_FLAG = -D_USE_BIG_SITE
-- SP_DMRG_FLAG = -D_USE_SP_DMRG
-- IC_FLAG = -D_USE_IC
-- KSYMM_FLAG =
-- SG_FLAG = -D_USE_SG
-- COMPLEX_FLAG =
-- SINGLE_PREC_FLAG =
-- SCI_FLAG =
-- TBB_FLAG =
-- MPI_FLAG = -D_HAS_MPI
-- OMP_LIB = /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/lib/libiomp5.so
-- MKL_OMP_LIB_NAME = mkl_intel_thread
-- TBB_LIBS =
-- Configuring done
CMake Warning at CMakeLists.txt:455 (ADD_LIBRARY):
  Cannot generate a safe runtime search path for target block2 because files
  in some directories may conflict with libraries in implicit directories:

    runtime library [libmkl_intel_lp64.so] in /opt/intel/compilers_and_libraries_2020.2.254/linux/mkl/lib/intel64 may be hidden by files in:
      /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/lib
    runtime library [libmkl_core.so] in /opt/intel/compilers_and_libraries_2020.2.254/linux/mkl/lib/intel64 may be hidden by files in:
      /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/lib
    runtime library [libmkl_intel_thread.so] in /opt/intel/compilers_and_libraries_2020.2.254/linux/mkl/lib/intel64 may be hidden by files in:
      /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/lib

  Some of these libraries may not be found correctly.


-- Generating done
-- Build files have been written to: /global/cscratch1/sd/zhcui/.consulting/block2-preview/build
```

Which then compiles for a while until being killed by the operating system because `icpc` is consuming too many resources:

```
icpc: error #10106: Fatal error in /opt/intel/compilers_and_libraries_2020.2.254/linux/bin/intel64/mcpcom, terminated by kill signal 
```

## Using GNU

```
zhcui@cori02:/global/cscratch1/sd/zhcui/.consulting/block2-preview/build> CC=cc CXX=CC cmake .. -DBUILD_LIB=ON -DMPI=ON -DOMP_LIB=GNU -DLARGE_BOND=ON -DUSE_SG=ON -DUSE_BIG_SITE=ON
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Cray Programming Environment 2.7.10 C
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /opt/cray/pe/craype/2.7.10/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Cray Programming Environment 2.7.10 CXX
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/cray/pe/craype/2.7.10/bin/CC - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/bin/python3 (found version "3.9.7")
-- Found PythonLibs: /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/lib/libpython3.so (found suitable version "3.9.7", minimum required is "3.9")
-- PROJECT_NAME = block2
-- PYTHON_VERSION_MAJOR = 3
-- PYTHON_VERSION_MINOR = 9
-- PYTHON_LIBRARIES = /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/lib/libpython3.so
-- PYTHON_EXECUTABLE = /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/bin/python3
-- PYTHON_EXECUTABLE_HINT =
-- PYTHON_INCLUDE_DIRS = /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/include/python3.9
-- PYLIB_SUFFIX = .cpython-39-x86_64-linux-gnu.so
-- PYBIND_INCLUDE_DIRS = /global/homes/z/zhcui/.local/cori/3.9-anaconda-2021.11/lib/python3.9/site-packages/pybind11/include
-- BUILD_LIB = ON
-- Looking for sgemm_
-- Looking for sgemm_ - found
-- Found BLAS: implicitly linked
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Looking for cheev_
-- Looking for cheev_ - found
-- Found LAPACK: implicitly linked
-- Found MPI_C: /opt/cray/pe/craype/2.7.10/bin/cc (found version "3.1")
-- Found MPI_CXX: /opt/cray/pe/craype/2.7.10/bin/CC (found version "3.1")
-- Found MPI: TRUE (found version "3.1")
-- MPI_COMPILE_FLAGS =
-- MPI_LINK_FLAGS =
-- MPI_INCLUDE_PATH =
-- MPI_CXX_LIBRARIES =
-- OPT_FLAG = -O3;-funroll-loops;-fopenmp;-Werror;-Werror=return-type;-Wno-psabi;-fvisibility=hidden;-Wno-error=attributes;-Wno-attributes
-- BOND_FLAG = -D_LARGE_BOND
-- MKL_FLAG =
-- CORE_FLAG = -D_USE_CORE
-- DMRG_FLAG = -D_USE_DMRG
-- BIG_SITE_FLAG = -D_USE_BIG_SITE
-- SP_DMRG_FLAG = -D_USE_SP_DMRG
-- IC_FLAG = -D_USE_IC
-- KSYMM_FLAG =
-- SG_FLAG = -D_USE_SG
-- COMPLEX_FLAG =
-- SINGLE_PREC_FLAG =
-- SCI_FLAG =
-- TBB_FLAG =
-- MPI_FLAG = -D_HAS_MPI
-- OMP_LIB = /global/common/software/nersc/cori-2022q1/sw/python/3.9-anaconda-2021.11/lib/libgomp.so
-- MKL_OMP_LIB_NAME = mkl_gnu_thread
-- TBB_LIBS =
-- Configuring done
-- Generating done
-- Build files have been written to: /global/cscratch1/sd/zhcui/.consulting/block2-preview/build
```
When trying to compile with `make` I get:
```
/global/cscratch1/sd/zhcui/.consulting/block2-preview/src/instantiation/core/../../core/csr_matrix_functions.hpp:759:21: error: no match for 'operator!=' (operand types are 'std::complex<double>' and 'int')
```

**Comment by hczhai** [comment]
Hi, thanks for the detailed information. I think both compilers should be okay, and I have the following suggestions:

(1) It seems that the ``icpc`` version 2019 (2020) has some internal issue which causes the memory problem. You can try a newer ``icpc`` compiler with version 2021 or 2022. I tried ``icpc 2021`` in MacOS (with 8G memory) and there is no memory issue.

(2) In the second part for ``gcc`` you did not use the cmake option ``-DUSE_MKL=ON`` (but in the first part for ``icpc`` you used the cmake option ``-DUSE_MKL=ON``). If you add the option ``-DUSE_MKL=ON`` then there should be no error messages. Building without ``-DUSE_MKL=ON`` will generate inefficient code. In case of you still want to build this inefficient version, you can update to the most recent version of the source code of ``block2``, and recompile, and the error will disappear.

**Comment by JBlaschke** [comment]
Hi @hczhai thanks for the information. 


The latest version of the Intel compiler that is available on Cori is Intel 19.1.3.20200925. Is there any way we can turn scale back the resource footprint used by icpc? Btw, an OOM error looks different, so it might also be the OS killing a process for using too many threads. TBH I haven't seen this error before, but then again I don't use the intel compiler all that much.

I tried with the `gcc` but also enabling `-DUSE_MKL=ON` but that causes the following error:
```
zhcui@cori02:/global/cscratch1/sd/zhcui/.consulting/block2-preview/build> CC=cc CXX=CC cmake .. -DUSE_MKL=ON -DBUILD_LIB=ON -DMPI=ON -DOMP_LIB=GNU -DLARGE_BOND=ON -DUSE_SG=ON -DUSE_B
IG_SITE=ON
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Cray Programming Environment 2.7.10 C
...
-- MKL_INCLUDE_DIR = MKL_INCLUDE_DIR-NOTFOUND
...
-- Configuring done
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
/global/cscratch1/sd/zhcui/.consulting/block2-preview/MKL_INCLUDE_DIR
   used as include directory in directory /global/cscratch1/sd/zhcui/.consulting/block2-preview
   used as include directory in directory /global/cscratch1/sd/zhcui/.consulting/block2-preview

-- Generating done
```
So I manually set the `-DMKL_INCLUDE_DIR=<path to MKL provided by the intel module>` which seems to have done the trick.

**Comment by JBlaschke** [comment]
For the record, earlier versions of MKL can cause issues with  missing libstdc++ symbols, eg:
```
In [1]: import block2
---------------------------------------------------------------------------
ImportError Traceback (most recent call last)
<ipython-input-1-07912f7ba129> in <module>
----> 1 import block2

ImportError: /global/u2/z/zhcui/program/block2/build_test/block2.cpython-39-x86_64-linux-gnu.so: undefined symbol: _ZTISt13runtime_error 
```

Using the latest available version of MKL (by setting `-DMKL_INCLUDE_DIR`) solved this issue.

I think we can work with this @hczhai  thank you. Closing the issue.
